compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1835,6 +1836,56 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    pub(super) fn collect_unused_for_coerce_return_ty(
+        &self,
+        errors: &[traits::FulfillmentError<'tcx>],
+    ) {
+        for fillment_error in &*errors {
+            let obligation = &fillment_error.obligation;
+            let body_id = obligation.cause.body_id;
+            let span = obligation.cause.span;
+
+            let Some(body_id) = self.tcx.hir().maybe_body_owned_by(
+                self.tcx.typeck_root_def_id(body_id.to_def_id()).expect_local(),
+            ) else {
+                continue;
+            };
+
+            let Some(mut diag) =
+                self.tcx.sess.diagnostic().steal_diagnostic(span, StashKey::MaybeForgetReturn)
+            else {
+                continue;
+            };
+
+            let body = self.tcx.hir().body(body_id);
+            let root_obligation = &fillment_error.root_obligation;
+            if let Some((_, fn_decl, _)) = self.get_fn_decl(body.value.hir_id)
+                && let hir::FnRetTy::Return(_ty) = fn_decl.output
+                && let Some(fn_sig) = self.body_fn_sig()
+                && let ExprBindingObligation(_, _, hir_id, ..) = root_obligation.cause.code() {
+                    for (_, node) in self.tcx.hir().parent_iter(*hir_id) {
```

Comment:
Semi should always be the direct parent of the `hir_id` of the binding obligation, I think?

Try `hir().get_parent()` instead of having a `for` loop over the whole parent iter?

Reply:
I'm unsure what depth we need to call `get_parent()`, it's maybe 1, 2, 3..

Reply:
Ugh, nevermind, it may be >1.